src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -650,6 +654,84 @@ def process_declarations(self, parent, nodes, arg_list):
+        information is used to determine the precision of the variable (as
+        supplied to the Symbol constructor).
+
+        :param symbol_table: the SymbolTable associated with the code \
+                             being processed.
+        :type symbol_table: :py:class:`psyclone.psyGen.SymbolTable`
+        :param type_spec: the fparser2 parse tree of the type specification.
+        :type type_spec: :py:class:`fparser.two.Fortran2003.Intrinsic_Type_Spec
+
+        :returns: the precision associated with the type specification.
+        :rtype: int or :py:class:`psyclone.psyGen.Symbol.Precision` or \
+                :py:class:`psyclone.psyGen.PrecisionSymbol` or NoneType
+
+        '''
+        if not isinstance(type_spec.items[1], Fortran2003.Kind_Selector):
+            return None
+        kind_selector = type_spec.items[1]
+
+        intrinsics = walk_ast(kind_selector.items,
+                              [Fortran2003.Intrinsic_Function_Reference])
+        if intrinsics and \
+           isinstance(intrinsics[0].items[0],
+                      Fortran2003.Intrinsic_Name) and \
+                      str(intrinsics[0].items[0]).lower() == "kind":
+            # We have kind=KIND(literal). items[1] is an
+            # Actual_Arg_Spec_List with the first entry being the
+            # literal.
+            literal = intrinsics[0].items[1].items[0]
+            if isinstance(literal,
+                          Fortran2003.Real_Literal_Constant):
+                if "d0" in str(literal).lower():
+                    kind_symbol = Symbol.Precision.DOUBLE
+                else:
+                    kind_symbol = Symbol.Precision.SINGLE
+            elif isinstance(literal,
+                            Fortran2003.Int_Literal_Constant):
+                kind_symbol = Symbol.Precision.SINGLE
+            else:
+                raise NotImplementedError(
+                    "Only names and (real and integer) literals "
+                    "are supported as KIND specifiers but found: "
+                    "{0}".format(str(kind_selector)))
+        else:
+            kind_names = walk_ast(kind_selector.items, [Fortran2003.Name])
+            if not kind_names:
+                raise NotImplementedError(
+                    "Failed to find valid Name in Fortran Kind "
+                    "Selector: '{0}'".format(str(kind_selector)))
+            # We have kind=var
+            kind_name = str(kind_names[0])
+            try:
+                kind_symbol = symbol_table.lookup(kind_name)
+                if not isinstance(kind_symbol, PrecisionSymbol):
+                    # The Symbol table contains an entry for the kind
+                    # parameter but it has not yet been identified as such. We
+                    # therefore replace the existing entry with a
+                    # PrecisionSymbol.
+                    new_symbol = PrecisionSymbol(kind_name)
```

Comment:
OK, so I don't think we should have a PrecisionSymbol. I think we should just use a Symbol. I understand that the precision variable will be declared in languages other than Fortran but I think that is right. It is actually just an integer parameter which can be used in other contexts as well. For example:
```
subroutine test()
  integer, parameter :: r_def = 4
  real(kind=r_def) :: a
  a = r_def
end subroutine
```
If we end up agreeing on this point then it simplifies this PR quite a bit as we no longer need the remove() method in the SymbolTable and no longer need to distinguish between symbols.

Reply:
Oh yes, I hadn't thought about such a symbol being used in other contexts. I'll change it back to being a `Symbol`.